eclair-core/src/main/scala/fr/acinq/eclair/Features.scala

```diff
@@ -38,6 +38,13 @@ trait Feature {
   def mandatory: Int
   def optional: Int = mandatory + 1
 
+  /** If true, feature should be sent in init messages. */
+  def init: Boolean
+  /** If true, feature should be sent in node announcements. */
+  def nodeAnnouncement: Boolean
+  /** If true, feature should be set in invoices. */
+  def invoice: Boolean
+
```

Comment:
> Interesting approach, but how will you deal with C/C-/C+ with a boolean? Is this why you have left those for later?

Yeah it's because it's unclear how these will ever be used in the spec (and they're currently unused).
So we'll deal with it if it ever becomes a thing.

> Btw, shouldn't the new option_payment_metadata be an I+?

That was discussed on the spec PR, at first it was 9+ but then it was changed because there's no reason not to allow it to be optional and it's a very good way to measure adoption before actually relying on it (and making it mandatory).

> why does the spec use 9 to identify invoice feature bits and not, say, V?

IIRC that's because features are encoded as `f` in invoices, and in bech32 `f` has the value `9`. I believe Rusty chose that a while ago when doing these categories.

Reply:
I switched to using traits in https://github.com/ACINQ/eclair/pull/2063/commits/4c39fd36e1f78f7efdfbdcc533dab6069174043f

I tried going further, and propagating the types to `Features`, but it becomes a mess quickly.
The main reason it's hard to properly handle is that we want to accept external data that doesn't comply with those type constraints (we don't want to be too strict and reject `init` or `node_announcement` that include a feature that we can simply ignore).